sentinelhub/constants.py

```diff
@@ -111,7 +112,41 @@ class _OrbitDirection(Enum):
     BOTH = 'both'
 
 
-class DataSource(Enum):
+class DataSourceMeta(EnumMeta):
+    """
+    Meta class used to add custom (BYOC) datasources to DataSource enumerator.
+
+    BYOC stands for Bring Your Own Data. For more details see: https://docs.sentinel-hub.com/api/latest/#/API/byoc
+    """
+    def __call__(cls, collection_id, *args, **kwargs):
+        """
+        This is executed whenever DataSource('something') is called.
+
+        The method raises a ValueError if the 'something' does not match the format expected for collection id.
+        """
+        collection_id_pattern = '.{8}-.{4}-.{4}-.{4}-.{12}'
+        if not re.compile(collection_id_pattern).match(collection_id):
+            raise ValueError("Given collection id does not match the expected format {}".format(collection_id_pattern))
+
+        def _custom_datasource_name(collection_id):
```

Comment:
In general defining a function inside a method is not very nice code practice. I recommend moving this to a static method of either `DataSourceMeta` or `DataSource` class.